### PR TITLE
a how-to guide about nested columns and arrays

### LINKED ISSUE #155: Document nested columns and arrays
The related PR is https://github.com/singularity-data/risingwave/pull/2434

Reference: https://cloud.google.com/bigquery/docs/nested-repeated#:~:text=To%20create%20a%20column%20with,to%20REPEATED%20in%20the%20schema.

### PR DESCRIPTION
**Pull request by hengm3467**
## Info
- **Description**: 
[ What's changed? ]

- **Related code PR**: 
https://github.com/risingwavelabs/risingwave/pull/2434

- **Related doc issue**: 
Resolves #155 

- **Notes**: 
Preview this page here: https://nested-columns-heng.d2fbku9n2b6wde.amplifyapp.com/docs/latest/nested-columns/

## Before merging
  - [ ] (For version-specific PR) I have selected the corresponding software version in **Milestone** and linked the related doc issue to this PR in **Development**.
  - [ ] I have acquired the approval from the owner (and optionally the reviewers) of the code PR and at least one tech writer (`bernscode`, `CharlieSYH`, `emile-00`, & `hengm3467`). 
  - [ ] I have checked the doc site preview, and the updated parts look good. <details><summary>How?</summary>image</details>

## Published doc pages
  [ After merging this PR, edit the description to include the links to the updated doc pages here. For example, https://www.risingwave.dev/docs/latest/intro/. ]

### CODE REVIEW

**Comment by CharlieSYH** [comment]
Preview: https://nested-columns-heng.d2fbku9n2b6wde.amplifyapp.com/docs/latest/nested-columns/

File: docs/guides/nested-columns.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            'FAST TAXI', 
+            '2030-12-31', 
+            'DAVID WANG'
+        )
+        );
+```
+
+## Access nested data
+
+To access data nested under a column, wrap the column with parenthesis and use the dot operator to specify the nested column. For example, to access the `id` column under `trip` in the `taxi_trips` schema, use `(trip).id`. For a complete example that defines and accesses nested columns, see [Fast Twitter events processing](../tutorials/fast-twitter-events-processing.md).
```
Comment (fuyufjh) [inline docs/guides/nested-columns.md]: Prefer to copy the examples here instead of reference

**Review by fuyufjh — COMMENTED** [review]
Generally LGTM.

1. As a doc under the "how-to guide" category, I think the tutorial of `Array` can also be covered in this article. (like [this](https://cloud.google.com/bigquery/docs/nested-repeated))
2. Need another doc under the SQL syntax folder for `Struct` data type.

File: docs/guides/nested-columns.md (hunk 1 above)
Comment (hengm3467) [reply docs/guides/nested-columns.md]: @fuyufjh As a data type, has `array` been fully supported yet? I tried creating a table with a field in `array` type, but got the message that the feature is not implemented yet. 

The `struct` data type is already described on the "Data types" page. I can include a link on that page to the current article.

Comment (CharlieSYH) [reply docs/guides/nested-columns.md]: Change `id` to `trip_id`.
For example, to access the `id` column under `trip` in the `taxi_trips` schema, use `(trip).id`.
⬇️
For example, to access the `trip_id` column under `trip` in the `taxi_trips` schema, use `(trip).trip_id`.

Comment (CharlieSYH) [inline docs/guides/nested-columns.md]: Suggestion:
`wrap the column with parenthesis`
⬇️
`wrap the column in parentheses`

File: sidebars.js [hunk 2]
```diff
@@ -91,7 +91,14 @@ const sidebars = {
         {type: 'ref', 
         label: 'Use time window functions', 
         id: 'sql/functions-operators/sql-function-time-window'
-        }
+        },
+
+        {
+          type: 'doc',
+          id: 'guides/nested-columns',
+          label: 'Define and access nested columns',
```
Comment (CharlieSYH) [inline sidebars.js:122]: We might also need to update the wording in [Data types - struct](https://www.risingwave.dev/docs/latest/sql-data-types/) and add a link to this detailed instruction.
image

Comment (hengm3467) [reply sidebars.js:122]: Thanks for the reminder. Let me add that.

File: docs/guides/nested-columns.md (hunk 1 above)
Comment (fuyufjh) [reply docs/guides/nested-columns.md]: > @fuyufjh As a data type, has `array` been fully supported yet? I tried creating a table with a field in `array` type, but got the message that the feature is not implemented yet.
> 
> The `struct` data type is already described on the "Data types" page. I can include a link on that page to the current article.

Yes, Array should be supported. Please open an issue for that 😇

Comment (fuyufjh) [reply docs/guides/nested-columns.md]: https://github.com/risingwavelabs/risingwave/blob/9d2e1d77faa5b71502e22d27705d2f5734e95c7f/e2e_test/streaming/project_set.slt#L4-L11

**Comment by hengm3467** [comment]
@fuyufjh I added instructions about specifying and accessing arrays. Please take a look and let me know if any description is incorrect or can be improved. Thanks!

**Review by fuyufjh — APPROVED** [review]
LGTM!

File: docs/guides/nested-columns-arrays.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
+---
+id: nested-columns-arrays
+title: Specify nested columns and arrays
+description: Specify nested columns and arrays
+slug: /nested-columns-arrays
+---
+This guide describes how to create a schema that contains nested columns and arrays (repeated columns), how to add values to such columns, and how to access nested and repeated data in RisingWave.
```
Comment (CharlieSYH) [inline docs/guides/nested-columns-arrays.md]: Omitting the last two "how to" can make the sentence sounds less repetitive.

File: docs/guides/nested-columns-arrays.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+To add values to nested columns, enclose the nested data with parentheses ('()') in the SQL statement. 
+
+To add values to an array, in the SQL statement, use `ARRAY` to indicate that this is an array, and then enclose the data in the array with square brackets ('[]'). For example, `ARRAY ['ABCD1234', 'ABCD1235', 'ABCD1236', 'ABCD1237']`.
+
+
+## Access nested data and data in an array
+
+To access data nested under a column, enclose the column with parentheses and use the dot operator to specify the nested column. For example, to access the `initial_charge` column under `fare` in the `trip` schema, use `(fare).initial_charge`.
+
+To access data in an array, use the `ARRAY_COLUMN[RELATIVE_POSITION]` syntax. Relative postions start from 1. For example, to access `ABCD1234`, which is the first object in the `trip_id` array, we can specify `trip_id[1]`.
```
Comment (CharlieSYH) [inline docs/guides/nested-columns-arrays.md]: Relative _postions_ start -> Relative **_positions_** start

Comment (CharlieSYH) [inline docs/guides/nested-columns-arrays.md]: "For example, to access `ABCD1234`, _which is_ the first object":
'which is' should go.

File: docs/guides/nested-columns-arrays.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+To create a column with nested data, set the data type of the column to `STRUCT`, and use angle brackets to enclose nested columns. Columns nested under the same column can be in different data types. A nested column can contain nested data. For more information about the `STRUCT` data type, see [Data types](../sql/sql-data-types.md).
+
+
+To create a column with repeated data (that is, an array), append [] to the data type of the column when you define the schema. For example, you can use `trip_id VARCHAR[]` to create an array that stores trip IDs.
+
+
+## Add values to nested columns and arrays
+
+To add values to nested columns, enclose the nested data with parentheses ('()') in the SQL statement.
```
Comment (CharlieSYH) [inline docs/guides/nested-columns-arrays.md]: I think the way we present the symbols of ('()') and ('[]') can be a little bit distracting:
image
image

"append [] to the data type of the column" in the previous paragraph looks good. Or maybe we can use the code format if it conforms to the style guide.

**Review by CharlieSYH — APPROVED** [review]
Added a few comments. The rest looks great!

File: docs/guides/nested-columns-arrays.md (hunk 5 above)
Comment (hengm3467) [reply docs/guides/nested-columns-arrays.md]: Noted. Just entering the brackets or parentheses (without quotation marks) works.

**Pull request merged**